plugins/crd/crd_utils.go

```diff
@@ -45,12 +46,31 @@ func newCustomResourceDefinition(plugin *Plugin, FullName, Group, Version, Plura
 		GetOpenAPIDefinitions: v1.GetOpenAPIDefinitions,
 	})
 
-	plugin.Log.Infof("Dumping CRD: %+v\n", crd)
+	crdClient := plugin.apiclientset.ApiextensionsV1beta1().CustomResourceDefinitions()
 
-	_, cserr := plugin.apiclientset.ApiextensionsV1beta1().CustomResourceDefinitions().Create(crd)
-	if apierrors.IsAlreadyExists(cserr) {
-		return nil
+	// First check if the CRD already exists
+	oldCRD, err := crdClient.Get(crd.Name, metav1.GetOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		plugin.Log.Errorf("error getting CRD %s, type %s", crd.Name, crd.Spec.Names.Kind)
+		return err
+	}
+	if apierrors.IsNotFound(err) {
+		// If the CRD does not exist, try to create it
+		if _, err := crdClient.Create(crd); err != nil {
+			plugin.Log.Errorf("error creating CRD %s, type %s", crd.Name, crd.Spec.Names.Kind)
```

Comment:
Do we want to do a get check again here, in case we were racing another creator.  Consider the possible race situation:

1.   NSM1 does a Get, and discovers the CRD doesn't exist
2.  NSM2 does a Get, and discovers the CRD doesn't exist
3.  NSM1 Creates the CRD, successfully
4.  NSM2 Creates the CRD, gets an error because NSM1 created it already
...
I'm suggesting NSM2's next step would be to go back to do the Get again, to check if it simply lost the race.

Reply:
@edwarnicke Note that is for creating the CRD definition in the k8s API server itself, not for individual CRDs. This is only run at startup time. So while it's *possible* for a race to exist here, it's not very likely. Does this information change your opinion on the patch?

Also, thanks for the review!

Reply:
Oh, I got this was for creating of the definition of the CRD, not specific instances of the CRD.  :)

I was just curious if we wanted to go with a belt and suspenders approach here.   I agree race conditions are unlikely.  Maybe we just talk through what a failure mode looks like...

I imagine the error eventually takes down the NSM (pretty quickly) on the Node... it exits.  The container exits.  The Pod exits.  The DaemonSet spawns a new one.  The new one doesn't hit the race condition.  The NSM Abides.

So I think what you have here is likely fine :)

Reply:
Yeah... as I think through it more deeply, I think the failure I described in my last comment is workable.  So this is fine :)

Reply:
We could check the error if it is "Already exists", then exit with success, otherwise failed.